### ISSUE TITLE
change existing command mode function [FEATURE REQUEST]

### ISSUE THREAD

**Issue by treymerkley** [xahlee/xah-fly-keys#70]
I've found this weird quirk that if I toggle command mode in neotree, it prevents me from using my keybind to toggle neotree unless I reload my init file. Is there a way to turn off xah-fly-keys only in the neotree buffer to remove the possibility of causing this?

**Comment by DanLanglois** [comment]
'it prevents me from using my keybind to toggle neotree'

Note that this is quite vague. I don't know what 'keybind' you mean  -- bluntly, is this going to be a thing? By 'using my keybind', you probably mean hitting a key, but I'm staring at a vague and portentious bug report, and thinking that i am no longer a young man.. 

So okay, perhaps you mean that <F8>, is how you open (toggle) NeoTree. This key is usually used to enter the Windows Startup menu or the Advanced Boot Options, and the assumption, I guess, is that nobody needs that, so Neotree overrides it. Unfortunately, I think this conflicts with 'important' command/insert mode switch keys of Xah-Fly-keys:

`;; xah-fly-command-mode-activate (press 【<home>】 or 【F8】 or 【Alt+Space】 or 【menu】)`

And let's look here:

https://github.com/xahlee/xah-fly-keys/blob/master/xah-fly-keys.el
(hash cc8b3bc)

This is line 3617:
`    (define-key xah-fly-key-map (kbd "<f8>") 'xah-fly-command-mode-activate-no-hook)`

I think you can implement a fix by commenting this line out with a semicolon:
`;    (define-key xah-fly-key-map (kbd "<f8>") 'xah-fly-command-mode-activate-no-hook)`

What this does is eliminates 'our' interest in this key. Otherwise, my concern is that hitting <F8> puts you into command mode, which isn't what you want, you want insert mode. 

I am not a committer, but I think it would be fine to pull request this fix, if it tests out working for you.

**Comment by billywade** [comment]
@DanLanglois generally, I tend to avoid making issues so obscure as to extract soliloquies from helpful strangers, so let me put in my config so you have a better idea as to what I mean. 

```
(use-package xah-fly-keys
  :ensure t
  :config
  (xah-fly-keys-set-layout "qwerty")
  (xah-fly-command-mode-init)
  (interactive)
  (define-key xah-fly-key-map (kbd "`") 'neotree-toggle)
  :diminish xah-fly-keys)
```
the "`" key by default has no binding in command mode, so ostensibly nothing should be conflicting with it. Let me know if you need more from me!

**Comment by DanLanglois** [comment]
I see your point about the '`' key. 

There are some details that are relevant, here, which I am guessing, at some of them. My first theory had been that  by default some of xah-fly-key bindings conflict with neotree-mode keys. I still worry that the issue of one of  overriding mappings in neotree-mode, or such. 


Okay, first, do you also have something like this?

`(add-to-list 'load-path "/some/path/neotree")`

And this?

`(require 'neotree)`

I mean, for me this sort of thing is in my 'init.el', same place where I put this kind of config stuff:

`(xah-fly-keys-set-layout "qwerty") ; required`


Now went to do some testing, installed neotree, and put this in init.el:

`(global-set-key [f2] 'neotree-toggle)`

I think I'm getting normal functionality, success. 

So I pushed my luck and tried this:

`(global-set-key [`] 'neotree-toggle)`

Also works as expected for me, full normal functionality. I've tried toggling around from insert to command mode, and opening files and such.

Any luck w/this idea?

**Comment by billywade** [comment]
Yeah, so I have neotree installed in my init file as
```
(use-package neotree
  :ensure t
  :init
  (setq neo-theme 'arrow)
  (setq neo-theme 'icons)
  (setq neo-smart-open t)
  (setq-default neo-show-hidden-files t)
  )
```
And ```global-set-key``` does work, but I was kind of hoping to keep it limited to xah's command mode so as to avoid taking up a key that I might use in insert mode.

Thanks for helping me with this!

**Comment by xahlee** [comment]
sorry guys been busy. will look at sometimes this week.

**Comment by billywade** [comment]
That's fine! You do you. It looks like "`" is bound to other-frame in xfk, and it's overriding my configuration, so all I need to do is unset it first.

**Comment by DanLanglois** [comment]
'And global-set-key does work, but I was kind of hoping to keep it limited to xah's command mode so as to avoid taking up a key that I might use in insert mode.'

Ah -- well, there is only one mutable keymap, currently, but I've put in a pull request replacing it with two keymaps, one for insert mode and one for command mode. 

You would use this, in init.el:

(define-key xah-fly-manage-command-mode-map (kbd "`") 'neotree-toggle)

As they say 'it works on my machine'.

**Comment by billywade** [comment]
quick update, I tried commenting out the line in xah-fly-keys that sets it to other frame, but it instead has essentially the same behaviour, except it inserts a literal "`" character when in command mode, so at a guess there's something wrong with how user-defined functions work.

**Comment by DanLanglois** [comment]
what line? Let me translate: 'I tried commenting out some line that I think does this but this is probably a poor guess as to what it does, but maybe you can make a guess that is a poor guess, as to what line I commented out? Progress!!' Sorry, just, translating so you know what this means in geekspeak.

**Comment by DanLanglois** [comment]
+1

--------
'And global-set-key does work, but I was kind of hoping to keep it limited to xah's command mode so as to avoid taking up a key that I might use in insert mode.'

Ah -- well, there is only one mutable keymap, currently, but I've put in a pull request replacing it with two keymaps, one for insert mode and one for command mode.

You would use this, in init.el:

(define-key xah-fly-manage-command-mode-map (kbd "`") 'neotree-toggle)

As they say 'it works on my machine'. If you like:

https://github.com/DanLanglois/xah-fly-keys

**Comment by billywade** [comment]
So I found a _really_ hacky solution in directly modifying xah-fly-command-mode-init in xah-fly-keys.el. Is there anything that can be done to make this something configurable via config?

**Comment by gcentauri** [comment]
The best way I've found to modify the command mode function maps is to use hooks.  There is the `xah-fly-command-mode-activate-hook` which I use to set a few things whenever command mode is toggled.  The issue is that the `xah-fly-key-map` gets re-initialized in a function every time it is toggled on.  So using `(define-key ...` like you do works once, but next time command mode is toggled on that key definition is wiped out.  I wrote a function `xah-fly-keys-command-mode-on` to hold all the things i want to happen when it toggles on, for example:

```elisp
(defun xah-fly-keys-command-mode-on ()
    (global-hl-line-mode 1)
    (set-cursor-color "deep pink")
    (setq cursor-type 'box)
    (define-key xah-fly-key-map (kbd "n") 'xah-fly-keys-search-function)
    (define-key xah-fly-key-map (kbd "`") 'neotree-toggle))
```
`xah-fly-keys-search-function` checks for `helm-swoop` or Ivy's `swiper` to use instead of `isearch-forward`

There's an added benefit to defining keys in a hook like this, they keyboard layout is already translated so when I define `(kbd "n")` there, it uses qwerty "n" instead of me having to know I should set the dvorak "b" key so that it can get translated to qwerty "n'.
